base_layer/core/tests/helpers/sync.rs

```diff
@@ -149,44 +165,48 @@ pub fn delete_some_blocks_and_headers(
     }
-    let set_best_block = set_best_block.unwrap_or(false);
     let mut blocks: Vec<_> = blocks_with_anchor.to_vec();
-    blocks.reverse();
-    for i in 0..blocks.len() - 1 {
-        let mut txn = DbTransaction::new();
-        match instruction {
-            WhatToDelete::Blocks => {
-                txn.delete_block(*blocks[i].hash());
-                txn.delete_orphan(*blocks[i].hash());
-                if set_best_block {
-                    txn.set_best_block(
-                        blocks[i + 1].height(),
-                        blocks[i + 1].accumulated_data().hash,
-                        blocks[i + 1].accumulated_data().total_accumulated_difficulty,
-                        *node.blockchain_db.get_chain_metadata().unwrap().best_block(),
-                        blocks[i + 1].to_chain_header().timestamp(),
-                    );
+    match instruction {
+        WhatToDelete::BlocksAndHeaders => {
+            node.blockchain_db.rewind_to_height(blocks[0].height()).unwrap();
+            for block in &blocks[1..] {
+                if node.blockchain_db.block_exists(*block.hash()).unwrap() {
+                    let mut txn = DbTransaction::new();
+                    txn.delete_orphan(*block.hash());
+                    node.blockchain_db.write(txn).unwrap();
                 }
-            },
-            WhatToDelete::Headers => {
-                txn.delete_header(blocks[i].height());
-            },
-        }
-        node.blockchain_db.write(txn).unwrap();
-        // Note: Something is funny here... the block is deleted but the block exists in the db. This should be
-        //       investigated and fixed as it will enhance the tests. If we uncomment the following assertion, the
-        //       tests depending on this function will fail.
-        // match instruction {
-        //     WhatToDelete::Blocks => {
-        //         assert!(!node.blockchain_db.block_exists(*blocks[i].hash()).unwrap());
-        //     }
-        //     WhatToDelete::Headers => {}
-        // }
+                assert!(!node.blockchain_db.block_exists(*block.hash()).unwrap());
+            }
+        },
+        WhatToDelete::Blocks => {
+            let headers = blocks.iter().map(|b| b.to_chain_header()).collect::<Vec<_>>();
+            node.blockchain_db.rewind_to_height(blocks[0].height()).unwrap();
+            for block in &blocks[1..] {
+                if node.blockchain_db.block_exists(*block.hash()).unwrap() {
+                    let mut txn = DbTransaction::new();
+                    txn.delete_orphan(*block.hash());
+                    node.blockchain_db.write(txn).unwrap();
+                }
+                assert!(!node.blockchain_db.block_exists(*block.hash()).unwrap());
+            }
+            node.blockchain_db.insert_valid_headers(headers[1..].to_vec()).unwrap();
+            // Note: This seems funny, as inserting the headers back into the db will cause this test to fail
```

Comment:
```suggestion
            // Note: This seems funny, as inserting the headers back into the db will cause this test to fail
        //       This should be investigated and fixed as it will enhance the tests. If we uncomment the following assertion, the
        //       tests depending on this function will fail.
```